video2dataset/data_reader.py

```diff
@@ -1,30 +1,72 @@
 """classes and functions for downloading videos"""
 import requests
 import tempfile
+import time
 import yt_dlp
 
+from multiprocessing import get_context
 
-QUALITY = "360p"
+from timeout_decorator import timeout, TimeoutError
 
-# TODO: update this to be link in video2numpy (with retries)
-def handle_youtube(youtube_url):
+
+def get_fast_format(formats, dl_timeout):
+    @timeout(dl_timeout)
+    def check_speed(f):
+        url = f.get('url')
+        ntf, _ = handle_mp4_link(url)
+        with open(ntf.name, "rb") as f:
+            vid_bytes = f.read()
+        ntf.close()
+        return
+
+    format_id = None
+    for f in formats:
+        try:
+            check_speed(f)
+            format_id = f.get('format_id')
+            break
+        except TimeoutError as e:
+            pass
+
+    return format_id
+
+
+def handle_youtube(youtube_url, retries, timeout, video_height, video_width):
     """returns file and destination name from youtube url."""
-    ydl_opts = {}
+    # Probe download speed:
+    ydl_opts = {
+        "quiet": True,
+        "socket_timeout": timeout,
+        "external-download": "ffmpeg",
+        "external-downloader-args": "ffmpeg_i:-ss 0 -t 2", # download 2 seconds
+    }
     ydl = yt_dlp.YoutubeDL(ydl_opts)
     info = ydl.extract_info(youtube_url, download=False)
     formats = info.get("formats", None)
-    f = None
-    for f in formats:
-        if f.get("format_note", None) != QUALITY:
-            continue
-        break
+    filtered_formats = [f for f in formats if f['height'] is not None and f['height'] >= video_height and f['width'] >= video_width]
```

Comment:
let's see with use, but might be best to filter after not before

Reply:
after what?

Reply:
I mean first retrieving the possible video, then filtering

Reply:
well if we want 360 height do we ever want videos with <360 height?

Reply:
yes, we can always resize up